python-squid/src/main/java/org/sonar/python/semantic/SymbolTableBuilderVisitor.java

```diff
@@ -0,0 +1,345 @@
+      }
+      if (globalNames.contains(symbolName)) {
+        return rootScope().resolve(symbolName);
+      }
+      return parent.resolve(symbolName);
+    }
+
+    private SymbolImpl resolveNonlocal(String symbolName) {
+      Scope scope = parent;
+      while (scope.parent != null) {
+        Symbol symbol = scope.symbolsByName.get(symbolName);
+        if (symbol != null) {
+          return (SymbolImpl) symbol;
+        }
+        scope = scope.parent;
+      }
+      return null;
+    }
+
+    private Scope rootScope() {
+      Scope scope = this;
+      while (scope.parent != null) {
+        scope = scope.parent;
+      }
+      return scope;
+    }
+
+    private void addGlobalName(String name) {
+      globalNames.add(name);
+    }
+
+    private void addNonlocalName(String name) {
+      nonlocalNames.add(name);
+    }
+  }
+
+  private static class SymbolImpl implements Symbol {
+
+    private final String name;
+    private final AstNode scopeRootTree;
+    private final Set<AstNode> writeUsages = new HashSet<>();
+    private final Set<AstNode> readUsages = new HashSet<>();
+
+    private SymbolImpl(String name, AstNode scopeRootTree) {
+      this.name = name;
+      this.scopeRootTree = scopeRootTree;
+    }
+
+    @Override
+    public String name() {
+      return name;
+    }
+
+    @Override
+    public AstNode scopeTree() {
+      return scopeRootTree;
+    }
+
+    @Override
+    public Set<AstNode> writeUsages() {
+      return Collections.unmodifiableSet(writeUsages);
+    }
+
+    @Override
+    public Set<AstNode> readUsages() {
+      return Collections.unmodifiableSet(readUsages);
+    }
+
+    public void addWriteUsage(AstNode nameNode) {
+      writeUsages.add(nameNode);
+    }
+
+    public void addReadUsage(AstNode nameNode) {
+      readUsages.add(nameNode);
+    }
+  }
+
+  private class SecondPhaseVisitor extends ScopeVisitor {
+
+    @Override
+    public Set<AstNodeType> subscribedKinds() {
+      return ImmutableSet.of(
+        PythonGrammar.FUNCDEF,
+        PythonGrammar.CLASSDEF,
+        PythonGrammar.ATOM,
+        PythonGrammar.DOTTED_NAME);
+    }
+
+    @Override
+    public void visitNode(AstNode node) {
+      super.visitNode(node);
+      if (node.is(PythonGrammar.ATOM, PythonGrammar.DOTTED_NAME)) {
+        visitAtomOrDottedName(node);
+      }
+    }
+
+  }
+
+  private class ClassVariableAssignmentVisitor extends ScopeVisitor {
+
+    public ClassVariableAssignmentVisitor(AstNode classTree) {
+      enterScope(classTree);
+    }
+
+    @Override
+    public Set<AstNodeType> subscribedKinds() {
+      return ImmutableSet.of(
+        PythonGrammar.FUNCDEF,
+        PythonGrammar.CLASSDEF,
+        PythonGrammar.ATOM,
+        PythonGrammar.DOTTED_NAME);
+    }
+
+    @Override
+    public void visitNode(AstNode node) {
+      super.visitNode(node);
+      if (node.is(PythonGrammar.ATOM, PythonGrammar.DOTTED_NAME) && currentScopeRootTree().is(PythonGrammar.CLASSDEF)) {
```

Comment:
@pynicolas If I remove the fragment
`currentScopeRootTree().is(PythonGrammar.CLASSDEF)`
from this line, the unit test still works. So the code fragment above can be removed, as this class is private and only used in a "class scope" context.